### DISCUSSION TITLE
Unable to import external Vue 3 component

### DISCUSSION THREAD

**Discussion by Destaq** [nuxt/nuxt#16419] [Questions] [answered]
I'm working on adding graph functionality to my Nuxt 3 project, but can't get any of the existing Vue 3 solutions to work with Nuxt 3. After trying several libraries such as Chart.js and HighCharts to no avail, I settled on [Vue-ECharts](https://github.com/ecomfe/vue-echarts#readme), given that it supports both Vue 3 and Vue 2 + the Nuxt Composition API - shown in the docs.

However, after Googling how to integrate these such Vue 3 components in Nuxt, I've come up empty. Even after trying to import it is a custom plugin as most resources suggest, there's no examples for Nuxt 3 (only Nuxt 2), and these don't work with the current framework.

I have also tried to set `"type": "module"` in `package.json` (no change) and importing the component in the specific `.vue` file (same error message).

<details>
<summary>The error I get (in short, *the dependency* cannot use an import statement outside the module)</summary>

```
Nuxt CLI v3.0.0-27333388.c824905                                                                                                                     23:01:50
                                                                                                                                                     23:01:50
  > Local:    http://localhost:3000/ 
  > Network:  http://10.0.0.4:3000/

ℹ Vite warmed up in 237ms                                                                                                                            23:01:51
✔ Vite server built in 5638ms                                                                                                                        23:01:56
✔ Nitro built in 63 ms                                                                                                                               23:01:56
(node:93789) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)
/Users/---/Desktop/vue-echarts-nuxt3/node_modules/echarts/lib/core/echarts.js:26
import { __extends } from "tslib";
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1026:15)
    at Module._compile (node:internal/modules/cjs/loader:1061:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1149:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:190:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:195:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:331:24)
[vite dev] Error loading external "/Users/---/Desktop/vue-echarts-nuxt3/node_modules/echarts/lib/core/echarts.js".
  at file://./.nuxt/dist/server/server.mjs:2034:299  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:121758:3)
[vite dev] Error loading external "/Users/---/Desktop/vue-echarts-nuxt3/node_modules/echarts/lib/core/echarts.js".
  at file://./.nuxt/dist/server/server.mjs:2034:299  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:121758:3)
[vite dev] Error loading external "/Users/---/Desktop/vue-echarts-nuxt3/node_modules/echarts/lib/core/echarts.js".
  at file://./.nuxt/dist/server/server.mjs:2034:299  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:121758:3)
[vite dev] Error loading external "/Users/---/Desktop/vue-echarts-nuxt3/node_modules/echarts/lib/core/echarts.js".
  at file://./.nuxt/dist/server/server.mjs:2034:299  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:121758:3)
```
</details>

You can find a minimum reproduction here: [repo](https://github.com/Destaq/vue-echarts-nuxt3)

Any tips to get this library to work with Nuxt 3, or alternative charting libraries that will work with the current framework?

**Comment by surfer77** [comment]
This is how I got a chart plugin to work, maybe can be useful for you


```
/plugins/chart.client.js

import { defineNuxtPlugin } from "#app";

import Vue3ChartJs from "@j-t-mcc/vue3-chartjs";

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(Vue3ChartJs);
});
```

on any component you can use it like this 

```
 <vue3-chart-js
        v-if="isMounted"
        style="height: 100%; width: 100%"
        id="day_transfers"
        type="bar"
        :data="stats"
        :options="options"
        @before-render="beforeRenderLogic"
      ></vue3-chart-js>
```

**Comment by danielroe** [comment, marked as answer]
A couple of minor changes to your repo: https://stackblitz.com/edit/github-unqktr

**Reply by Destaq** [reply]
Thank you!

**Reply by quanming1** [reply]
hi,i wonder how did you install this nuxt3 project by  npm ?

**Reply by zengshine** [reply]
> A couple of minor changes to your repo: https://stackblitz.com/edit/github-unqktr

while using webpack as builder: https://stackblitz.com/edit/github-5qqbff-jovjya?file=nuxt.config.ts
something goes wrong: 
![image](https://user-images.githubusercontent.com/18556375/190963615-9a5575fb-9635-41f3-8205-1ebed1724089.png)

**Reply by zieru** [reply]
> A couple of minor changes to your repo: https://stackblitz.com/edit/github-unqktr
not working while building
![image](https://user-images.githubusercontent.com/1273033/201032433-b4eeafa3-2752-42e6-a687-e4abb6bf20d8.png)

**Reply by danielroe** [reply]
That's because this is a very outdated issue and we entered RC since then. We now do not install `nuxt3` but `nuxt`.

**Comment by benoitdemaegdt** [comment]
Hey there 👋 

In order to help futur readers, here is how I got [Highcharts](https://www.highcharts.com/) working with Nuxt3

1/ Install [Highcharts-Vue](https://github.com/highcharts/highcharts-vue)
```bash
npm install highcharts-vue
```

2/ create a nuxt3 plugin
```bash
mkdir plugins && touch plugins/highcharts.client.js
```
Copy/Paste this content inside this newly created file
```js
import HighchartsVue from 'highcharts-vue'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(HighchartsVue)
})
```

3/ create a component containing your chart
For instance you have a file `/components/Chart.vue`
```vue
<template>
  <highcharts :options="chartOptions"></highcharts>
</template>

<script>
export default {
  name: 'Chart',
  setup() {
    return {
      chartOptions: {
        series: [
          {
            data: [1, 2, 3], // sample data
          },
        ],
      },
    }
  },
}
</script>
```

4/ wrap this component call inside a ClientOnly tag
Example in `pages/index.vue`
```vue
<template>
  <ClientOnly>
    <Chart />
  </ClientOnly>
</template>
```

The chart should now display correctly 🎉

**Reply by EL-AKKAL** [reply]
saved my life after 4 days of trying to make it work , thanks !!

**Comment by DannyAndres** [comment]
I think this is not the complete solution or maybe I'm missing something. Destaq asked about adding vue-echarts which is composed by echarts module and vue-echarts module as well, I did not had any problem with echarts using [danielroe](https://github.com/danielroe) repo but if I try to import vue-echarts module inside a plugin like [benoitdemaegdt](https://github.com/benoitdemaegdt) did, simply does not work and it says "Failed to resolve component: v-chart". So basically I'm stuck with installing echarts tools but I cannot install the component to use them. Did you guys had a problem like that ?

**Reply by DannyAndres** [reply]
To reply myself I solved it by basically changing the line 

`nuxtApp.vueApp.use(VChart)`

to 

`nuxtApp.vueApp.component('v-chart', VChart);`

This way the component is now registered, the echart tools are imported as the solution's repo shows
